### PR TITLE
[UR][hip][opencl] Mark urKernelSuggestMaxCooperativeGroupCountExp as unsupported instead of returning misleading default value

### PR DESCRIPTION
**Pull request by GeorgeWeb**
intel/llvm: https://github.com/intel/llvm/pull/15369

### CODE REVIEW

**Review by frasercrmck — APPROVED** [review]
Makes sense to me

**Review by jchlanda — APPROVED** [review]

File: source/adapters/opencl/kernel.cpp [hunk 1]
```diff
@@ -358,11 +358,11 @@ UR_APIEXPORT ur_result_t UR_APICALL urKernelGetNativeHandle(
 UR_APIEXPORT ur_result_t UR_APICALL urKernelSuggestMaxCooperativeGroupCountExp(
     ur_kernel_handle_t hKernel, size_t localWorkSize,
     size_t dynamicSharedMemorySize, uint32_t *pGroupCountRet) {
-  (void)hKernel;
-  (void)localWorkSize;
-  (void)dynamicSharedMemorySize;
-  *pGroupCountRet = 1;
-  return UR_RESULT_SUCCESS;
+  std::ignore = hKernel;
```
Comment (aarongreig) [inline source/adapters/opencl/kernel.cpp:361]: we usually prefer `[[maybe_unused]]` in the cl adapter, or just not naming the parameters

Comment (GeorgeWeb) [reply source/adapters/opencl/kernel.cpp:361]: Yeah, me too to be honest. I wasn't sure whether that'll be acceptable as I mostly see `std::ignore`s around the codebase. Now that you've said, I'll change it to your preference.

Comment (GeorgeWeb) [reply source/adapters/opencl/kernel.cpp]: Updated.

**Review by aarongreig — APPROVED** [review]

**Comment by jinz2014** [comment]
@GeorgeWeb 
Could you please explain that it is not supported on the HIP backend ?

**Comment by jinz2014** [comment]
What about the following function for HIP ?

```
UR_APIEXPORT ur_result_t UR_APICALL urKernelSuggestMaxCooperativeGroupCountExp(
    ur_kernel_handle_t hKernel, ur_device_handle_t hDevice, uint32_t workDim,
    const size_t *pLocalWorkSize, size_t dynamicSharedMemorySize,
    uint32_t *pGroupCountRet) {
  UR_ASSERT(hKernel, UR_RESULT_ERROR_INVALID_KERNEL);

  std::ignore = hDevice;

  size_t localWorkSize = pLocalWorkSize[0];
  localWorkSize *= (workDim >= 2 ? pLocalWorkSize[1] : 1);
  localWorkSize *= (workDim == 3 ? pLocalWorkSize[2] : 1);

  // We need to set the active current device for this kernel explicitly here,
  // because the occupancy querying API does not take device parameter.
  ur_device_handle_t Device = hKernel->getProgram()->getDevice();
  ScopedContext Active(Device);
  try {
    // We need to calculate max num of work-groups using per-device semantics.

    int MaxNumActiveGroupsPerCU{0};
    UR_CHECK_ERROR(hipOccupancyMaxActiveBlocksPerMultiprocessor(
        &MaxNumActiveGroupsPerCU, hKernel->get(), localWorkSize,
        dynamicSharedMemorySize));
    detail::ur::assertion(MaxNumActiveGroupsPerCU >= 0);
    // Handle the case where we can't have all SMs active with at least 1 group
    // per SM. In that case, the device is still able to run 1 work-group, hence
    // we will manually check if it is possible with the available HW resources.
    if (MaxNumActiveGroupsPerCU == 0) {
      size_t MaxWorkGroupSize{};
      urKernelGetGroupInfo(
          hKernel, Device, UR_KERNEL_GROUP_INFO_WORK_GROUP_SIZE,
          sizeof(MaxWorkGroupSize), &MaxWorkGroupSize, nullptr);
      size_t MaxLocalSizeBytes{};
      urDeviceGetInfo(Device, UR_DEVICE_INFO_LOCAL_MEM_SIZE,
                      sizeof(MaxLocalSizeBytes), &MaxLocalSizeBytes, nullptr);
      if (localWorkSize > MaxWorkGroupSize ||
          dynamicSharedMemorySize > MaxLocalSizeBytes)
        *pGroupCountRet = 0;
      else
        *pGroupCountRet = 1;
    } else {
      // Multiply by the number of SMs (CUs = compute units) on the device in
      // order to retreive the total number of groups/blocks that can be
      // launched.
      *pGroupCountRet = Device->getNumComputeUnits() * MaxNumActiveGroupsPerCU;
    }
  } catch (ur_result_t Err) {
    return Err;
  }
  return UR_RESULT_SUCCESS;
}
```

**Comment by GeorgeWeb** [comment]
> Could you please explain that it is not supported on the HIP backend ?

Hi @jinz2014.
It isn't unsupported by the HIP runtime itself as the required API exists, but we just haven't implemented it in the HIP adapter of Unified Runtime yet. This is mostly due to that the functionality was requested for Cuda targets at the time.

You are more than welcome to open a Github issue for the team in case they may have the bandwidth for this or straight up a PR request for HIP with the implementation you suggested in your other comment.

Thank you for being proactive on the project!

**Pull request merged**